crates/noirc_evaluator/src/ssa/optim.rs

```diff
@@ -307,33 +240,60 @@ pub fn find_similar_mem_instruction(
         }
         _ => unreachable!("invalid non memory operator"),
     }
-    (ins_id, CseAction::Keep)
+
+    CseAction::Keep
 }
 
 pub fn propagate(ctx: &SsaContext, id: NodeId) -> NodeId {
     let mut result = id;
     if let Some(obj) = ctx.try_get_instruction(id) {
-        if obj.operator == node::Operation::Ass || obj.is_deleted {
-            result = obj.rhs;
+        if let Operation::Binary(node::Binary { operator: BinaryOp::Assign, rhs, .. }) =
+            &obj.operator
+        {
+            result = *rhs;
+        }
+
+        if let Some(replacement) = obj.replacement {
+            result = replacement;
         }
     }
     result
 }
 
 //common subexpression elimination, starting from the root
 pub fn cse(igen: &mut SsaContext, first_block: BlockId) -> Option<NodeId> {
-    let mut anchor = HashMap::new();
+    let mut anchor = Anchor::default();
     cse_tree(igen, first_block, &mut anchor)
 }
 
+/// A list of instructions with the same Operation variant, ordered by the order
+/// they appear in their respective blocks.
+#[derive(Default, Clone)]
+struct Anchor {
+    map: HashMap<Opcode, VecDeque<NodeId>>,
+}
+
+impl Anchor {
+    fn push_front(&mut self, op: &Operation, id: NodeId) {
+        let key = match op {
+            Operation::Store { array_id, .. } => Opcode::Load(*array_id),
+            _ => op.opcode(),
+        };
+        self.map.entry(key).or_insert_with(VecDeque::new).push_front(id);
+    }
+
+    fn get_all(&self, opcode: Opcode) -> Cow<VecDeque<NodeId>> {
```

Comment:
I don't see the benefit of doing this versus what was done; i.e getting a reference to the entry with `&anchor[&ins.operator]`

Reply:
This doesn't push empty VecDeques for each opcode anymore so the return type needed to be changed to represent a borrowed VecDeque or an owned (empty) VecDeque.